### ISSUE TITLE
NTP possible with BBA?

### ISSUE THREAD

**Issue by epochVHS** [emukidid/swiss-gc#472]
I know it's an enhancement with a very limited audience, but would it be possible to add a simple NTP sync to Swiss? Another blade in the Swiss army knife, if you will. Or is time sync limited to the IPL?

Could make it auto-sync on network initialization as well.

Apologies if this has come up before-I searched for a while but couldn't find anything about it.

**Comment by JinxDojo** [comment]
My Gamecube's battery must be going because the time is always getting more and more behind. NTP would be nice!

**Comment by ronny332** [comment]
My plan was to start with a FTP Server implementation as first addon for Swiss (if wanted for the main project, otherwise for myself as playground :blush:), but starting today, I got a NTP requester up and running. Currently it is standalone, but it should be easily to add it to the network initialization process.
The only unknown thing to me until now: can the internal clock be set :smile:?
![ntp_test](https://local.nn332.de/apps/files_sharing/publicpreview/yS83RT7TSxZXWWD?x=1680&y=977&a=true&file=ntp.png&scalingup=0)

**Comment by Extrems** [comment]
I recommend using `__SYS_SetRTC` and leaving the clock bias for timezone adjustment (usually only the bias is modified).

**Comment by ronny332** [comment]
Thanks for the advice :+1: . I will start tomorrow with testing the different ways to update the system clock. For now I have to write an alternative for net_gethostbyname(). The stub is in the gamecube libogc source, but the integration in the library just exists for the wii variant.

**Comment by Extrems** [comment]
There's no integration because it simply doesn't exist in our terribly old version of lwIP.

**Comment by ronny332** [comment]
Yeah, anyway, for me it's a kind of assignment. I haven't written C code from the basement upwards. Makes a lot of fun.

**Comment by Extrems** [comment]
It shouldn't be necessary for this purpose if we use the Cloudflare or Google NTP servers.

**Comment by ronny332** [comment]
I know ;-).

**Comment by ronny332** [comment]
Ok, done. "getipforhost()" works. I had a bit problems with domains, which send a whole set of load balancing devices (google, facebook, ...). Further, the Gamecube network environment seems not to have a possibility to read out the currently used DNS server, or I'm just blind. Now it should work (for now with a default/external one), more to come later ...

![getipforhost](https://local.nn332.de/apps/files_sharing/publicpreview/JpWEeE6K79C9p5D?x=1680&y=977&a=true&file=getipbyhost.png&scalingup=0)

**Comment by Extrems** [comment]
They are requested by the DHCP client, but they're only accessible through lwIP APIs... which aren't accessible.

**Comment by ronny332** [comment]
> I recommend using `__SYS_SetRTC` and leaving the clock bias for timezone adjustment (usually only the bias is modified).

`__SYS_GetRTC` is part of libogc2, right? `__SYS_SetRTC` on the other hand exists and works. I'm able to find it at the swiss object file in libogc.a, but not at the normally packed library.
One of my next steps are switching to your libog2 anyways, but for now I was using devkitpro without modifications.

**Comment by Extrems** [comment]
Here's the adjustment necessary: extremscorner/gamecube-examples@0c92030

Also note that our epoch is January 1, 2000, not January 1, 1970.

**Comment by ronny332** [comment]
Ok, I ran into issues ...
If have the unix timestamp transformed to a gc timestamp, set and read the values (`__SYS_GetRTC` and `__SYS_SetRTC`), both values are identical, even after a cold start of the cube. But the Calendar values and Swiss clock stays the same.
At first I thought, that setting the clock is not persistent, but that seems to be wrong. Before setting a newly fetched value from an NTP server the old value matches the new one.

![timestamps](https://local.nn332.de/apps/files_sharing/publicpreview/Pq58k5LeHTtR4Kq?x=1674&y=962&a=true&file=ntp_set.jpg&scalingup=0)

**Comment by Extrems** [comment]
What do you mean by staying the same?

**Comment by ronny332** [comment]
Hi, yes, the clock values of Swiss and the normal Gamecube Calendar are staying at the date, they were set by hand before.
As far as I can see there are no checks in the libogc2 code to prevend a wrong timestamp to be stored.
And the values should be fin. 679193325 + the time from 1.1.1970 to 1.1.2000 in seconds, combines to roughly the unix timestamp from the picture above.

**Comment by Extrems** [comment]
There's no such thing as a wrong timestamp.
You may fix your clock bias using `SYS_SetCounterBias`, but you may not fix others', as this is necessary to exploit game mechanics.

**Comment by ronny332** [comment]
Yes, indeed. It is related to the Gamecube I used. Another, also a PAL model (but this shouldn't matter) works fine. Setting the time to `7200` results in 1.1.2000 at 2:00am. I will open the first Gamecube, remove the RTC battery for a moment, and try to clear the SRAM.

**Comment by GyroDragona** [comment]
Has there been any progress with adding NTP clock setting option to swiss?

**Comment by ronny332** [comment]
Currently I'm working on different platforms, but the timestamp reading for NTP was working. Somebody has to implement the code from above. After there was no reply to my last response, I assumed nobody was interested in a further usage.

**Comment by GyroDragona** [comment]
I would love to see this implemented. The link seems to be dead. Is there a new place to get it from? Now that there is a cheap alternative in the ETH2GC instead of the OEM broadband adapter this could be a lot more relevant.

**Comment by Extrems** [comment]
Officially speaking this is still stalled at the lwIP upgrade. It seemed like it was about to happen a month ago, but I was suddenly distracted with other work.

**Comment by GyroDragona** [comment]
I think it would be a super useful feature that most common users would enjoy even if they don't know what it is or even that it is there working in the background. It is used in nearly every device that is able to get online to set the time automatically.
